src/shogun/classifier/NearestCentroid.cpp

```diff
@@ -54,51 +53,46 @@ namespace shogun{
 		{
 			data = distance->get_lhs();
 		}
-		int32_t num_vectors = data->get_num_vectors();
-		int32_t num_classes = ((CMulticlassLabels*) m_labels)->get_num_classes();
-		int32_t num_feats = ((CDenseFeatures<float64_t>*) data)->get_num_features();
-		SGMatrix<float64_t> centroids(num_feats,num_classes);
+
+		auto* multiclass_labels = m_labels->as<CMulticlassLabels>();
+		auto* dense_data = data->as<CDenseFeatures<float64_t>>();
+
+		int32_t num_classes = multiclass_labels->get_num_classes();
+		int32_t num_feats = dense_data->get_num_features();
+
+		SGMatrix<float64_t> centroids(num_feats, num_classes);
 		centroids.zero();
 
-		int64_t* num_per_class = new int64_t[num_classes];
-		for (int32_t i=0 ; i<num_classes ; i++)
-		{
-			num_per_class[i]=0;
-		}
+		SGVector<int64_t> num_per_class(num_classes);
+		num_per_class.zero();
 
-		for (int32_t idx=0 ; idx<num_vectors ; idx++)
+		auto iter_labels = multiclass_labels->get_int_labels().begin();
+		for(const auto& current : DotIterator(dense_data))
 		{
-			int32_t current_len;
-			bool current_free;
-			int32_t current_class = ((CMulticlassLabels*) m_labels)->get_label(idx);
-			float64_t* target = centroids.matrix + num_feats*current_class;
-			float64_t* current = ((CDenseFeatures<float64_t>*)data)->get_feature_vector(idx,current_len,current_free);
-			SGVector<float64_t>::add(target,1.0,target,1.0,current,current_len);
+			const auto current_class = *(iter_labels++);
+			auto target = centroids.get_column(current_class);
+			current.add(1, target);
 			num_per_class[current_class]++;
-			((CDenseFeatures<float64_t>*)data)->free_feature_vector(current, current_len, current_free);
 		}
 
-
 		for (int32_t i=0 ; i<num_classes ; i++)
 		{
-			float64_t* target = centroids.matrix + num_feats*i;
+			auto target = centroids.get_column(i);
 			int32_t total = num_per_class[i];
 			float64_t scale = 0;
 			if(total>1)
 				scale = 1.0/((float64_t)(total-1));
 			else
 				scale = 1.0/(float64_t)total;
 
-			SGVector<float64_t>::scale_vector(scale,target,num_feats);
+			target.scale(scale);
```

Comment:
I wonder whether we should offer a linalg method to do this.....Might be a bit faster if done in eigen matrix-wise
`scale_cols(SGMatrix<T> mat, SGVector<T> scalars)`

Reply:
I think a column/row manipulation api would be better than getting the row/column first, then operating on it. I found it a bit confusing that `get_column` returns an SGVector that maps to the matrix, but `get_row` copies the values (which makes sense since it's in column-major).

Reply:
It is confusing indeed. I think in fact that the `get_row` should be removed, or rather renamed to `copy_row`.
And then we can add a `linalg` call to operate on the columns

Reply:
`scale_cols(SGMatrix<T> mat, SGVector<T> scalars)` function would make things cleaner
Does implementing it require doing it in the different backends?
Should I also change `get_row` to `copy_row`? I think that's a better name.

Reply:
Just do it in the eigen3 backend for now. See how the other things are added. Btw the linalg thing is a big mess inside. Maybe we can think about simplifying some things too during gsoc.

If possible I would just delete the row getter as it encourages bad things. And then possibly refactor the code that uses it? How many places do call it?

Reply:
Oh and maybe name it just ‘scale’ as the signature tells the story already.